src/api/server.rs

```diff
@@ -0,0 +1,123 @@
+use super::{handler, schema::build_schema};
+
+use async_graphql::{
+    http::{playground_source, GraphQLPlaygroundConfig},
+    QueryBuilder,
+};
+use async_graphql_warp::{graphql_subscription, GQLResponse};
+use std::convert::Infallible;
+use std::net::SocketAddr;
+use tokio::sync::oneshot;
+use tokio::sync::oneshot::{Receiver, Sender};
+use warp::filters::BoxedFilter;
+use warp::{http::Response, Filter, Reply};
+
+pub struct Server {
+    /// Address for the API server to bind on
+    address: SocketAddr,
+
+    /// Transmission channel to trigger closure of a running API server
+    trigger_cancel: Sender<()>,
+
+    /// Receiver signal to cancel running API server
+    cancel_signal: Option<Receiver<()>>,
+}
+
+impl Server {
+    /// Returns a new API Server
+    pub fn new(address: SocketAddr) -> Server {
+        let (trigger_cancel, cancel_signal) = oneshot::channel::<()>();
+
+        Server {
+            address,
+            trigger_cancel,
+            cancel_signal: Some(cancel_signal),
+        }
+    }
+
+    /// String representation of the bound IP address
+    pub fn ip(&self) -> String {
+        self.address.ip().to_string()
+    }
+
+    /// String representation of the bound port
+    pub fn port(&self) -> String {
+        self.address.port().to_string()
+    }
+
+    /// Stops the running API server
+    pub fn stop(self) {
+        let _ = self.trigger_cancel.send(());
+    }
+
+    /// Run the API server
+    pub async fn run(mut self) -> Self {
+        let rx = self
+            .cancel_signal
+            .take()
+            .expect("Run can only be called once");
+
+        let (_, server) =
+            warp::serve(make_routes()).bind_with_graceful_shutdown(self.address, async move {
+                let _ = rx.await;
+            });
+
+        tokio::spawn(server);
+
+        self
+    }
+}
+
+/// Builds Warp routes, to be served
+fn make_routes() -> BoxedFilter<(impl Reply,)> {
+    // Health route
+    let health_route = warp::path("health").and_then(handler::health);
+
+    // Build the GraphQL schema
+    let schema = build_schema().finish();
+
+    // GraphQL POST handler
+    let graphql_post = async_graphql_warp::graphql(schema.clone()).and_then(
+        |(schema, builder): (_, QueryBuilder)| async move {
+            let resp = builder.execute(&schema).await;
+            Ok::<_, Infallible>(GQLResponse::from(resp))
+        },
+    );
+
+    // GraphQL playground. Also sets up /graphql as an endpoint for both queries + subscriptions
+    let graphql_playground = warp::path("playground").map(|| {
+        Response::builder()
+            .header("content-type", "text/html")
+            .body(playground_source(
+                GraphQLPlaygroundConfig::new("/graphql").subscription_endpoint("/graphql"),
```

Comment:
I was wondering if we should use a path that describes the data you can get at that endpoint, vs. the technology used to query that data? (f.e. `/metrics` vs `/graphql`?). But I'm not even sure that's necessarily better, and it's a minor nit regardless.

Reply:
The current intention is to use GraphQL as the fuller interface for the API (and `vector top`), which is currently:

- Health / heartbeats
- Metrics (in progress)
- Reading / validating `Cargo.toml` (in progress)

The roadmap isn't super clear atm, so I'm not sure how much this will evolve, but it feels likely that metrics alone probably won't encompass the full UI <-> Vector comms.

What _is_ likely though, I think, is having multiple interfaces into the same API, which as regular HTTP endpoints for the same thing. Therefore, I think `/graphql` + `/<other_interfaces>` is a good place to start.

Reply:
Yeah, that makes sense 👍